pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -0,0 +1,797 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scbuildstmt
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/paramparse"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgnotice"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scerrors"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catid"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+func alterTableAlterPrimaryKey(
+	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, t *tree.AlterTableAlterPrimaryKey,
+) {
+	// Panic on certain forbidden `ALTER PRIMARY KEY` cases (e.g. one of
+	// the new primary key column is a virtual column). See the comments
+	// for a full list of preconditions we check.
+	checkForEarlyExit(b, tbl, t)
+
+	// Nothing to do if the requested new pk is the same as the old one.
+	if isNewPrimaryKeySameAsOldPrimaryKey(b, tbl, t) {
+		return
+	}
+
+	// TODO (xiang): This section contains all fall-back cases and need to
+	// 							 be removed to fully support `ALTER PRIMARY KEY`.
+	fallBackIfConcurrentSchemaChange(b, tbl.TableID)
+	fallBackIfRequestToBeSharded(t)
+	fallBackIfSecondaryIndexExists(b, tbl.TableID)
+	fallBackIfRegionalByRowTable(b, tbl.TableID)
+	fallBackIfDescColInRowLevelTTLTables(b, tbl.TableID, t)
+
+	// Retrieve old primary index and its name elements.
+	oldPrimaryIndexElem, newPrimaryIndexElem := getPrimaryIndexes(b, tbl.TableID)
+	if newPrimaryIndexElem != nil {
+		// TODO (Xiang): some other DDL stmt preceded this `ALTER PRIMARY KEY` and
+		//               thus a new primary index has already been created. We'd like
+		//               to support this use case one day
+		//               (e.g. `ALTER TABLE t ADD COLUMN ..., ALTER PRIMARY KEY ...;`).
+		//               Note that such scenarios should be caught above in
+		//               `fallBackIfConcurrentSchemaChange` and an unimplemented error
+		//               should be returned, so, here we panic with an programming error.
```

Comment:
nit you don't need to wrap this like this, I think it's harder to read and annoying to maintain.